distributed/protocol/serialize.py

```diff
@@ -135,6 +135,41 @@ def serialize(x, serializers=None, on_error="message", context=None):
     if isinstance(x, Serialized):
         return x.header, x.frames
 
+    if isinstance(x, (dict, list, set, tuple)) and len(x) <= 5:
+        if isinstance(x, dict):
+            t = []
+            for k, v in x.items():
+                h, f = serialize(
+                    v, serializers=serializers, on_error=on_error, context=context
+                )
+                h["key"] = pickle.dumps(k)
+                t.append((h, f))
+        else:
+            t = [
+                serialize(
+                    i, serializers=serializers, on_error=on_error, context=context
+                )
+                for i in x
+            ]
+
+        frames = []
+        lengths = []
+        for _header, _frames in t:
+            frames.extend(_frames)
+            length = len(_frames)
+            if "is-collection" in _header and _header["is-collection"] is True:
+                length = sum(_header["frame-lengths"])
```

Comment:
This is needed because of recursion, if the object we're serializing is a `tuple` in a `tuple` and the innermost one has multiple frames, we have to make sure we account for all frames it has.

Reply:
But shouldn't all of those frames be in `_frames`?  I would expect that ...

```python
len(_frames) == sum(_header["frame-lengths"])
```

Reply:
More generally, the frames returned by a serialize call should always be entirely flattened, so we shouldn't have to think about nesting at all.

Reply:
I suppose you're right. Yesterday I spent some time debugging this and there was a case where that wasn't working, but removing this no longer fails, really not sure what was going on. Nevertheless, better this way. :)